### PR TITLE
When invoking class constructor ensure class is initialized

### LINKED ISSUE #1748: static constructors throw NRE if having struct fields and invoked via TypeInitializer
The following code throws NullReferenceException on .NET Core (doesn't throw on .NET Framework).
It this a bug?

```cs
using System;

struct S { }

class A
{
    static TimeSpan x = default; // throws NRE
    // FYI, generated il is: ldsflda, initobj
}

class B
{
    static S x = default; // throws NRE
}

class C
{
    static S x;
    static C() => x = default; // throws NRE
}

class D
{
    // no exception if static fiealds are primitives or reference types
    // generated il is: ldnull/ldc.i4, stsfld
    static string x = null;
    static int y = 0;
}

static class Program
{
    static void Main()
    {
        Console.WriteLine("----");
        try { typeof(A).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("A"); } // A
        try { typeof(B).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("B"); } // B
        try { typeof(C).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("C"); } // C
        try { typeof(D).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("D"); } // (no excpetion)

        Console.WriteLine("----");
        A a = new A();
        B b = new B();
        C c = new C();
        try { typeof(A).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("A"); } // A
        try { typeof(B).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("B"); } // B
        try { typeof(C).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("C"); } // (no excpetion)
        try { typeof(D).TypeInitializer.Invoke(null, null); } catch { Console.WriteLine("D"); } // (no excpetion)
    }
}
```

### PR DESCRIPTION
**Pull request by davidwrighton**
When invoking the class constructor method via reflection invoke, ensure that the class constructor is run via the standard run class constructor pathway before manually invoking the class constructor

This ensure that all of the various data structures associated with the class constructor are initialized such as valuetype statics.

Fixes #1748

### CODE REVIEW

File: src/libraries/System.Reflection/tests/ConstructorInfoTests.cs [hunk 1]
```diff
@@ -68,6 +68,25 @@ public void Invoke_StaticConstructor_NullObject_NullParameters()
             Assert.Null(obj);
         }
 
+        [Fact]
+        public void Invoke_StaticConstructorMultipleTimes()
+        {
+            ConstructorInfo[] constructors = GetConstructors(typeof(ClassWithStaticConstructorThatIsCalledMultipleTimesViaReflection));
```
Comment (danmosemsft) [inline src/libraries/System.Reflection/tests/ConstructorInfoTests.cs:74]: BTW this test library doesn't run against .NET Framework, so you'd have to paste this into a test program if you wanted to be sure NETFX behavior was the same.

**Comment by jkotas** [comment]
Does this create easy path to mutate read-only statics (by accident)?

We have disallowed setting of read-only fields via reflection in .NET Core. This is a more subtle variant of the same. I think it may be better to fix this by disallowing static cctor invocation via reflection.

**Comment by jkotas** [comment]
> I think it may be better to fix this by disallowing static cctor invocation via reflection.

Or to allow it just once.

**Comment by davidwrighton** [comment]
This creates all sorts of questionable paths. But they aren't new. @danmosemsft I've verified that this is the behavior of Netfx 4.8.

I'd be pleased to convert an attempt invoke the static ctor via this mechanism into a simple call to RunClassConstructor instead of actually, you know, repeatedly invoking the class constructor directly. The current checked in scheme is clearly not safe, as it doesn't behave reliably, and can skip critical aspects of cctor invocation.

**Comment by jkotas** [comment]
> I'd be pleased to convert an attempt invoke the static ctor via this mechanism into a simple call to RunClassConstructor

+1

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs [hunk 2]
```diff
@@ -280,6 +286,14 @@ internal void ThrowNoInvokeException()
             // check basic method consistency. This call will throw if there are problems in the target/method relationship
             CheckConsistency(obj);
 
+            if ((invocationFlags & INVOCATION_FLAGS.INVOCATION_FLAGS_RUN_CLASS_CONSTRUCTOR) != 0)
+            {
+                // Run the class constructor through the class constructor mechanism instead of the Invoke path.
+                // This avoids allowing mutation of readonly static fields, and initializes the type correctly.
+                RuntimeHelpers.RunClassConstructor(DeclaringType!.TypeHandle);
```
Comment (jkotas) [inline src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs:293]: Can we get here for module constructors? I believe `DeclaringType` is going to be null for them.

**Review by jkotas — APPROVED** [review]
LGTM!

File: src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs (hunk 2 above)
Comment (davidwrighton) [reply src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeConstructorInfo.cs]: Sigh, we can. I'll update the code to handle them by running the module constructor api.

**Pull request merged**